lms/djangoapps/course_blocks/transformers/user_partitions.py

```diff
@@ -79,16 +88,25 @@ def transform_block_filters(self, usage_info, block_structure):
             return [block_structure.create_universal_filter()]
 
         user_groups = get_user_partition_groups(usage_info.course_key, user_partitions, user, 'id')
-        group_access_filter = block_structure.create_removal_filter(
-            lambda block_key: not (
-                has_access(user, 'staff', block_key) or
-                block_structure.get_transformer_block_field(block_key, self, 'merged_group_access').check_group_access(
-                    user_groups
-                )
+        for block_key in block_structure.topological_traversal():
+            transformer_block_field = block_structure.get_transformer_block_field(
+                block_key, self, 'merged_group_access'
             )
-        )
-
-        result_list.append(group_access_filter)
+            access_denying_partition_id = transformer_block_field.get_access_denying_partition(
+                user_groups
+            )
+            access_denying_partition = get_partition_from_id(user_partitions, access_denying_partition_id)
+            if not has_access(user, 'staff', block_key) and access_denying_partition:
+                user_group = user_groups[access_denying_partition_id]
+                allowed_groups = transformer_block_field._access[access_denying_partition_id]
+                access_denied_message = access_denying_partition.access_denied_message(block_key, user, user_group, allowed_groups)
```

Comment:
What will happen if `access_denied_message` returns `None` (which is what the base `UserPartition` returns).

I think what we want to have happen is for those blocks to not be included in the response (that is, they should behave like the v1 behavior). So, maybe we need to do some filtering in this transformer as well?

Reply:
well i made the access_denied_filter transformer remove blocks that don't have both a filter and a message.
although i realize now that for api v2 that transformer won't be included, so I guess I want to duplicate the same filtering here

Reply:
I mean, maybe you just always include that filter, and have the v1 argument make it more or less aggressive?

Reply:
well i guess part of the reason you wanted to have that separate transformer in the first place was so it would be easier to deprecate v1.  
but I am now handling the case where access_denied_message is None in user_partitions.py